### PR TITLE
Pr ci integration in git-sensor

### PR DESCRIPTION
**Pull request by manish-agrawal-ai**
1. Webhook API in git-sensor
    2. Database changes and scripts
    3. Github parsing of webhook event and save/update in DB

### CODE REVIEW

File: pkg/RepoManages.go [hunk 1]
```diff
@@ -387,15 +403,71 @@ func (impl RepoManagerImpl) FetchChanges(pipelineMaterialId int, from string, to
...
+	err := json.Unmarshal([]byte(pipelineMaterial.CommitHistory), &commits)
 	if err != nil {
 		return nil, err
 	}
 	response.Commits = commits
 	return response, nil
+}
+
+
+func (impl RepoManagerImpl) FetchGitCommitsForPrType(pipelineMaterial *sql.CiPipelineMaterial) (*git.MaterialChangeResp, error) {
```
Comment (pghildiyal) [inline pkg/RepoManages.go:415]: should it be PR or should it be webhook? question is - is this generic enough to handle any webhook for CI trigger or only PR based trigger

File: pkg/RepoManages.go [hunk 2]
```diff
@@ -387,15 +403,71 @@ func (impl RepoManagerImpl) FetchChanges(pipelineMaterialId int, from string, to
...
+
+	webhookMappings, err := impl.webhookEventRepository.GetCiPipelineMaterialPrWebhookMapping(pipelineMaterialId)
+	if err != nil {
+		impl.logger.Errorw("error in getting webhook mapping for pipelineId ", "id", pipelineMaterialId, "errMsg", err)
+		return nil, err
+	}
+
+	if len(webhookMappings) == 0 {
+		impl.logger.Infow("no webhook mapping for ci pipeline, pipelineId", "pipelineId", pipelineMaterialId)
+		return nil, err
```
Comment (pghildiyal) [inline pkg/RepoManages.go:428]: err will be nil in this case

File: pkg/RepoManages.go [hunk 3]
```diff
@@ -387,15 +403,71 @@ func (impl RepoManagerImpl) FetchChanges(pipelineMaterialId int, from string, to
...
+
+	var prWebhookEventIds []int
+	for _, webhookMapping := range webhookMappings {
+		prWebhookEventIds = append(prWebhookEventIds, webhookMapping.PrWebhookDataId)
+	}
+
+	prWebhookEvents, err := impl.webhookEventRepository.GetOpenPrEventDataByIds(prWebhookEventIds, 15)
+	if err != nil {
+		impl.logger.Errorw("error in getting webhook for ids ", "ids", prWebhookEventIds, "errMsg", err)
+		return nil, err
```
Comment (pghildiyal) [inline pkg/RepoManages.go:439]: err will be nil in this case, is that the intention? then we should send nil explicity

File: pkg/RepoManages.go [hunk 4]
```diff
@@ -387,15 +403,71 @@ func (impl RepoManagerImpl) FetchChanges(pipelineMaterialId int, from string, to
...
+	}
+
+	if len(prWebhookEvents) == 0 {
+		impl.logger.Infow("no webhooks for ci pipeline, pipelineId", "pipelineId", pipelineMaterialId)
+		return nil, err
+	}
+
+	commits := make([]*git.GitCommit, 0)
```
Comment (pghildiyal) [inline pkg/RepoManages.go:447]: recommended way to do it is
var commits []*git.GitCommit

File: pkg/RepoManages.go [hunk 5]
```diff
@@ -499,3 +571,29 @@ func (impl RepoManagerImpl) RefreshGitMaterial(req *git.RefreshGitMaterialReques
...
+
+func (impl RepoManagerImpl) GetPrEventDataById(id int) (*git.PrData, error) {
+	webhookPRDataEvent, err := impl.webhookEventRepository.GetPrEventDataById(id)
+
+	if err != nil {
+		impl.logger.Errorw("error in getting pr event for Id ", "Id", id, "err", err)
+		return nil, err
+	}
+
+	PrData := &git.PrData{
```
Comment (pghildiyal) [inline pkg/RepoManages.go:584]: we are create PrData object from webhookPRDataEvent above also, should extract out this transformation and reuse it

File: pkg/git/Bean.go [hunk 6]
```diff
@@ -56,13 +61,39 @@ type GitCommit struct {
...
+}
+
+
 type CommitMetadataRequest struct {
 	PipelineMaterialId int    `json:"pipelineMaterialId"`
 	GitHash            string `json:"gitHash"`
 	GitTag             string `json:"gitTag"`
 }
 
+type LatestCommitMetadataRequest struct {
```
Comment (pghildiyal) [inline pkg/git/Bean.go:88]: I dont see any field which restricts it to Latest, so we should not have Latest in the name of the struct.. also cant we combine it with the previous struct i.e; CommitMetadataRequest, ll there be some logical inconsistency or purpoose will not be clear if we combine them?

File: pkg/git/GitOperationService.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	FetchAndGetCommitInfo(pipelineMaterialId int, gitHash string) (*GitCommit, error)
+	GetLatestCommitForBranch(pipelineMaterialId int, branchName string) (*GitCommit, error)
+}
+
+type GitOperationServiceImpl struct {
+	logger                       *zap.SugaredLogger
+	materialRepository           sql.MaterialRepository
+	ciPipelineMaterialRepository sql.CiPipelineMaterialRepository
+	locker                       *internal.RepositoryLocker
+	gitRepositoryManager         RepositoryManager
```
Comment (pghildiyal) [inline pkg/git/GitOperationService.go:38]: We dont add words like manager to Repository any specific reason to do it here

File: pkg/git/WebhookEventService.go [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+	}
+
+	if len(materials) == 0 {
+		impl.logger.Infow("no materials found skipping.")
+		return nil
+	}
+
+	for _, material := range materials {
+		ciPipelineMaterials := material.CiPipelineMaterials
+		if len(ciPipelineMaterials) == 0 {
```
Comment (pghildiyal) [inline pkg/git/WebhookEventService.go:122]: redundant check..

File: pkg/git/GitOperationService.go [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+	"fmt"
+	"github.com/devtron-labs/git-sensor/internal"
+	"github.com/devtron-labs/git-sensor/internal/sql"
+	"go.uber.org/zap"
+	_ "gopkg.in/robfig/cron.v3"
+	"gopkg.in/src-d/go-git.v4"
+)
+
+type GitOperationService interface {
+	FetchAndGetCommitInfo(pipelineMaterialId int, gitHash string) (*GitCommit, error)
```
Comment (pghildiyal) [inline pkg/git/GitOperationService.go:29]: Fetch is same as Get, name should be FetchCommitInfo

File: internal/sql/CiPipelineMaterial.go [hunk 10]
```diff
@@ -38,6 +38,11 @@ type CiPipelineMaterial struct {
 	ErrorMsg      string     `sql:"error_msg,notnull"`
 }
 
+type PullRequestSourceTypeValue struct {
```
Comment (pghildiyal) [inline internal/sql/CiPipelineMaterial.go:41]: This should not be defined in repo

File: pkg/git/RepositoryManager.go [hunk 11]
```diff
@@ -206,6 +208,31 @@ func (impl RepositoryManagerImpl) GetCommitMetadata(checkoutPath, commitHash str
 	return gitCommit, nil
 }
 
+
+func (impl RepositoryManagerImpl) GetCommitMetadataForRepository(repository *git.Repository, commitHash string) (*GitCommit, error) {
```
Comment (pghildiyal) [inline pkg/git/RepositoryManager.go:212]: can we combine this and GetLatestCommitForBranch

File: internal/sql/CiPipelineMaterial.go (hunk 10 above)
Comment (manish-agrawal-ai) [reply internal/sql/CiPipelineMaterial.go:41]: done

File: pkg/RepoManages.go (hunk 1 above)
Comment (manish-agrawal-ai) [reply pkg/RepoManages.go:415]: Done

File: pkg/RepoManages.go (hunk 2 above)
Comment (manish-agrawal-ai) [reply pkg/RepoManages.go:428]: done

File: pkg/RepoManages.go (hunk 3 above)
Comment (manish-agrawal-ai) [reply pkg/RepoManages.go:439]: done

File: pkg/RepoManages.go (hunk 4 above)
Comment (manish-agrawal-ai) [reply pkg/RepoManages.go:447]: done

File: pkg/RepoManages.go (hunk 5 above)
Comment (manish-agrawal-ai) [reply pkg/RepoManages.go:584]: done

File: pkg/git/Bean.go (hunk 6 above)
Comment (manish-agrawal-ai) [reply pkg/git/Bean.go:88]: ok.
CommitMetadataRequest is having gitHash in the request object.
LatestCommitMetadataRequest is having branchName in the request object.

Implementation is different for these two. The later one will fetch the last commit info for that branch name. 

But ofcourse we can merge this.

File: pkg/git/GitOperationService.go (hunk 9 above)
Comment (manish-agrawal-ai) [reply pkg/git/GitOperationService.go:29]: Not needed now. Have deleted the code

File: pkg/git/GitOperationService.go (hunk 7 above)
Comment (manish-agrawal-ai) [reply pkg/git/GitOperationService.go:38]: This is old class. (RepositoryManager)
Not related to this PR.

File: pkg/git/Bean.go (hunk 6 above)
Comment (manish-agrawal-ai) [reply pkg/git/Bean.go:88]: merged

**Pull request closed without merging**